src/modules/appointments/appointments.service.ts

```diff
@@ -0,0 +1,139 @@
+import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { CreateAppointmentInput } from './dto/create-appointment.input';
+import { Appointment } from './entities/appointment.entity';
+
+@Injectable()
+export class AppointmentsService {
+  constructor(
+    @InjectRepository(Appointment)
+    private appointmentsRepository: Repository<Appointment>,
+  ) {}
+  async create(
+    createAppointmentInput: CreateAppointmentInput,
+  ): Promise<Appointment> {
+    try {
+      //* Getting all appointments in this room
+      const appointmentsInThisRoom = await this.findManyWithOptions({
+        where: { roomId: createAppointmentInput.roomId },
```

Comment:
You should perform here a check to verify it the slot is available by using "find where roomId = x, scheduleDate = input.date, scheduleTime >= input.date.time and scheduleTime <= input.date.time + input.duration.

You will get an intersection (overlap) with existing appointments, and you can throw an exception instantly.

In case of no overlap - you can freely create a new appointment.